admiral/pkg/clusters/envoyfilter.go

```diff
@@ -0,0 +1,153 @@
+package clusters
+
+import (
+	"fmt"
+	"github.com/gogo/protobuf/types"
+	v1 "github.com/istio-ecosystem/admiral/admiral/pkg/apis/admiral/v1"
+	"github.com/istio-ecosystem/admiral/admiral/pkg/controller/admiral"
+	"github.com/istio-ecosystem/admiral/admiral/pkg/controller/common"
+	log "github.com/sirupsen/logrus"
+	"istio.io/api/networking/v1alpha3"
+	networking "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strings"
+)
+
+func createOrUpdateEnvoyFilter( rc *RemoteController, routingPolicy *v1.RoutingPolicy, eventType admiral.EventType, workloadIdentityKey string, admiralCache *AdmiralCache) (*networking.EnvoyFilter, error) {
+
+	workloadSelectors := make(map[string]string)
+	workloadSelectors[common.GetWorkloadIdentifier()] = workloadIdentityKey
+	workloadSelectors[common.GetEnvKey()] = common.GetRoutingPolicyEnv(routingPolicy)
+
+	envoyfilterSpec := constructEnvoyFilterStruct(routingPolicy, workloadSelectors)
+
+	selectorLabelsSha, err := common.GetSha1(workloadIdentityKey+common.GetRoutingPolicyEnv(routingPolicy))
+	if err != nil {
+		log.Error("Error ocurred while computing workload Labels sha1")
+		return nil, err
+	}
+	envoyFilterName := fmt.Sprintf("%s-dynamicrouting-%s-%s", strings.ToLower(routingPolicy.Spec.Plugin), selectorLabelsSha, "1.10")
+	envoyfilter := &networking.EnvoyFilter{
+		TypeMeta: metaV1.TypeMeta{
+			Kind:       "EnvoyFilter",
+			APIVersion: "networking.istio.io/v1alpha3",
+		},
+		ObjectMeta: metaV1.ObjectMeta{
+			Name:      envoyFilterName,
+			Namespace: "istio-system",
+		},
+		Spec: *envoyfilterSpec,
+	}
+
+	admiralCache.RoutingPolicyFilterCache.Put(workloadIdentityKey+common.GetRoutingPolicyEnv(routingPolicy), rc.ClusterID, envoyFilterName)
+	var filter *networking.EnvoyFilter
+	//get the envoyfilter if it exists. If it exists, update it. Otherwise create it.
+	if eventType == admiral.Add || eventType == admiral.Update {
+		filter, err = rc.RoutingPolicyController.IstioClient.NetworkingV1alpha3().EnvoyFilters("istio-system").Get(envoyFilterName, metaV1.GetOptions{})
```

Comment:
are we expecting an error to be returned if the filter doesn't already exist?
because, if the value of filer is nil, in that case, there is a chance of a nil pointer failure on line number 55

Reply:
If filter does not exist, this API call results into an error. Next line is checking for `error` and if it is not `nil` then we create the filter if it does not exist.